### PR TITLE
Fix types & refactor createStyled

### PR DESCRIPTION
**Pull request by gregberge**
I fixed some types and refactor `createStyled`. I am not very far to have a modular thing.

### CODE REVIEW

File: packages/styled-components/src/createStyled.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+): CreateXStyled<StyleGeneratorProps<TGen>> => {
+  const config = generator
+    ? {
+        shouldForwardProp: createShouldForwardProp(generator),
+      }
+    : {}
+  return ((component: Parameters<typeof scStyled>[0]) => {
+    const baseStyled = scStyled(component)
+    return getCreateStyle(
+      config ? baseStyled.withConfig(config) : baseStyled,
```
Comment (agriffis) [inline packages/styled-components/src/createStyled.ts:80]: config will always be truthy here, maybe the conditional above should be `: null` or `: undefined`?

File: packages/styled-components/src/createStyled.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    defaultValidatorFn: (prop: string | number | symbol) => boolean,
+    elementToBeCreated?: ElementType,
+  ) => {
+    if (string(prop) && propSet.has(prop)) {
+      return false
+    }
+    if (typeof elementToBeCreated === 'string') {
+      // We must test elementToBeCreated so we can pass through props for <x.div
+      // as={Component} />. However elementToBeCreated isn't available until
+      // styled-components 5.2.4 or 6, and in the meantime will be undefined.
```
Comment (agriffis) [inline packages/styled-components/src/createStyled.ts:59]: FYI this fix was released in styled-components 5.3.0 — both of my guesses were wrong!

File: packages/styled-components/src/createStyled.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      generator,
+    )
+  }) as CreateXStyled<StyleGeneratorProps<TGen>>
+}
+
+export const createStyled = <TGen extends StyleGenerator>(
+  generator: TGen,
+): CreateXStyled<StyleGeneratorProps<TGen>> => {
+  const styled = createBaseStyled()
+  const xstyled = createBaseStyled(generator)
```
Comment (agriffis) [inline packages/styled-components/src/createStyled.ts:90]: This is a nice way to do this :+1:

File: packages/styled-components/src/createStyled.ts (hunk 1 above)
Comment (gregberge) [reply packages/styled-components/src/createStyled.ts:80]: I will change it in the next PR.

File: packages/styled-components/src/createStyled.ts (hunk 2 above)
Comment (gregberge) [reply packages/styled-components/src/createStyled.ts:59]: I will change it too.

**Pull request merged**